avm/src/vm/event.rs

```diff
@@ -204,25 +208,31 @@ impl HandlerFragment {
     &mut self,
     mut hand_mem: Arc<HandlerMemory>,
     instrs: &Vec<Instruction>,
-  ) -> Arc<HandlerMemory> {
+  ) -> VMResult<Arc<HandlerMemory>> {
     task::block_in_place(move || {
-      instrs.iter().for_each(|i| {
-        if let OpcodeFn::Cpu(func) = i.opcode.fun {
-          //eprintln!("{} {} {} {}", i.opcode._name, i.args[0], i.args[1], i.args[2]);
-          let event = func(i.args.as_slice(), &mut hand_mem);
-          if event.is_some() {
-            let event_tx = EVENT_TX.get().unwrap();
-            let event_sent = event_tx.send(event.unwrap());
-            if event_sent.is_err() {
-              eprintln!("Event transmission error");
-              std::process::exit(2);
+      instrs
+        .iter()
+        .map(|i| {
+          if let OpcodeFn::Cpu(func) = i.opcode.fun {
+            //eprintln!("{} {} {} {}", i.opcode._name, i.args[0], i.args[1], i.args[2]);
+            let event = func(i.args.as_slice(), &mut hand_mem);
+            if let Some(event) = event? {
+              let event_tx = EVENT_TX.get().unwrap();
+              let event_sent = event_tx.send(event);
+              if event_sent.is_err() {
+                eprintln!("Event transmission error");
+                std::process::exit(2);
+              }
             }
+            Ok(())
+          } else {
```

Comment:
This whole branch is known (to us) to never happen because of the way the opcode fragments work. It would be nice to be able to assert this with the type during the fragment generation, perhaps having another enum for the "fragment type" wrapping around the vec and only allowing the particular opcode function enum we care about within each of these outer enums and being able to avoid this looping check here?

Not something for this PR, but it does bother me that we have to define an error path here that will never be triggered and that we have to pay a price doing these useless checks.

Reply:
I think there's a way to unify the function types of the different opcode kinds, so then we can entirely get rid of these checks and only use the additional information for building up the instruction fragments. I imagine it'd be a struct with 2 fields, and running them would be much faster.

The new opcode defs would look like:
```rust
pub enum OpcodeKind {
  CPU,
  IO,
  UnpredCPU,
}
pub struct Opcode {
  name: &'static str,
  kind: OpcodeKind,
  exec: fn(&[i64], Arc<HandlerMemory>) -> Box<Pin<VMResult<Arc<HandlerMemory>>>>,
}
```
so execution can be done with purely an iterator (simplified):
```rust
handler.instructions
  .iter()
  .map(|fragment| {
    fragment
      .iter()
      .map(|(op, instrs)| exec(&instrs, hmem.fork()))
      .collect::<UnorderedFutures<VMResult<Vec<Arc<HandlerMemory>>>>>>()
  })
  .collect::<OrderedFutures<VMResult<Vec<Arc<HandlerMemory>>>>>()
```

Reply:
So the CPU-bound opcodes can return an event payload, they don't return the HandlerMemory.

However, I think that weirdness is used only by the `emitop` opcode, so if we can refactor things to let event emission be done specially by that opcode and not require checking the return value, that'd be great. But what I'd prefer is that CPU opcodes still have a special return type, it's just that it's now `void` and we don't have to do any conditional checking on every single CPU opcode call, which is probably a significant perf impact.